src/workerd/io/worker.c++

```diff
@@ -448,37 +448,105 @@ public:
+  }
 
-    return (timePoint - kj::UNIX_EPOCH) / kj::MILLISECONDS;
+  void setChannel(Worker::Isolate::InspectorChannelImpl& channel) {
+    auto lockedState = state.lockExclusive();
+    lockedState->channel = channel;
   }
 
-  // Nothing else. We ignore everything the inspector tells us, because we only care about the
-  // devtools inspector protocol, which is handled separately.
+  void resetChannel() {
+    auto lockedState = state.lockExclusive();
+    lockedState->channel = {};
+  }
 
-  void setInspectorTimerInfo(kj::Timer& timer, kj::Duration timerOffset) {
-    // Helper for attachInspector().
-    inspectorTimerInfo = InspectorTimerInfo { timer, timerOffset, getCurrentThreadId() };
+  void runMessageLoopOnPause(int contextGroupId) override {
+    // This method is called by v8 when a breakpoint or debugger statement is hit. This method
+    // processes debugger messages until `Debugger.resume()` is called, when v8 then calls
+    // `quitMessageLoopOnPause()`.
+    //
+    // This method is ultimately called from the `InspectorChannelImpl` and the isolate lock is
+    // held when this method is called.
+    if (isMultiTenantProcess()) {
+      // TODO: breakpoints are not supported for the multi-tenant case.
+      return;
+    }
+
+    {
+      auto lockedState = state.lockExclusive();
+      KJ_IF_MAYBE(channel, lockedState->channel) {
+        pauseIncomingMessages(*channel);
+      }
+    }
+
+    runMessageLoop = true;
+    do {
+      auto lockedState = state.lockExclusive();
+      // Processing each message with a lock/release cycle is expensive, but
+      // we are debugging and processing a low volume of messages. Dropping the
+      // lock allows us to adapt behaviour if the lock is lost.
+      KJ_IF_MAYBE(channel, lockedState->channel) {
+        if (!dispatchOneMessageDuringPause(*channel)) {
+          break;
+        }
+      } else {
+        // Our channel is missing or has gone down. Stop waiting for messages
+        // and resume execution.
+        break;
+      }
+    } while (runMessageLoop);
+
+    {
+      auto lockedState = state.lockExclusive();
+      KJ_IF_MAYBE(channel, lockedState->channel) {
+        resumeIncomingMessages(*channel);
+      }
+    }
+  }
+
+  void quitMessageLoopOnPause() override {
+    // This method is called by v8 to resume execution after a breakpoint is hit.
+    runMessageLoop = false;
   }
 
 private:
+  static void pauseIncomingMessages(Worker::Isolate::InspectorChannelImpl& channel);
+  static void resumeIncomingMessages(Worker::Isolate::InspectorChannelImpl& channel);
+  static bool dispatchOneMessageDuringPause(Worker::Isolate::InspectorChannelImpl& channel);
+
   struct InspectorTimerInfo {
     kj::Timer& timer;
     kj::Duration timerOffset;
     uint64_t threadId;
   };
 
-  kj::Maybe<InspectorTimerInfo> inspectorTimerInfo;
-  // The timer and offset for the inspector-serving thread.
+  volatile bool runMessageLoop;
```

Comment:
Is this really volatile? It looks like it will only be accessed in methods that run under isolate lock.

Reply:
Oops, done, that's a very early debugging bandaid.